src/cmd/services/m3coordinator/downsample/downsampler.go

```diff
@@ -0,0 +1,756 @@
+		return newAggregatorResult{}, err
+	}
+
+	var (
+		storageFlushConcurrency = defaultStorageFlushConcurrency
+		rulesStore              = o.RulesKVStore
+		clockOpts               = o.ClockOptions
+		instrumentOpts          = o.InstrumentOptions
+	)
+	if o.StorageFlushConcurrency > 0 {
+		storageFlushConcurrency = storageFlushConcurrency
+	}
+
+	// Configure rules options.
+	tagEncoderPool := serialize.NewTagEncoderPool(o.TagEncoderOptions,
+		o.TagEncoderPoolOptions)
+	tagEncoderPool.Init()
+
+	tagDecoderPool := serialize.NewTagDecoderPool(o.TagDecoderOptions,
+		o.TagDecoderPoolOptions)
+	tagDecoderPool.Init()
+
+	sortedTagIteratorPool := newEncodedTagsIteratorPool(tagDecoderPool,
+		o.TagDecoderPoolOptions)
+	sortedTagIteratorPool.Init()
+
+	sortedTagIteratorFn := func(tagPairs []byte) id.SortedTagIterator {
+		it := sortedTagIteratorPool.Get()
+		it.Reset(tagPairs)
+		return it
+	}
+
+	tagsFilterOptions := filters.TagsFilterOptions{
+		NameTagKey: metricNameTagName,
+		NameAndTagsFn: func(id []byte) ([]byte, []byte, error) {
+			return encodedTagsNameAndTags(id, sortedTagIteratorPool)
+		},
+		SortedTagIteratorFn: sortedTagIteratorFn,
+	}
+
+	isRollupIDFn := func(name []byte, tags []byte) bool {
+		return isRollupID(tags, sortedTagIteratorPool)
+	}
+
+	newRollupIDProviderPool := newRollupIDProviderPool(tagEncoderPool,
+		o.TagEncoderPoolOptions)
+	newRollupIDProviderPool.Init()
+
+	newRollupIDFn := func(name []byte, tagPairs []id.TagPair) []byte {
+		rollupIDProvider := newRollupIDProviderPool.Get()
+		id, err := rollupIDProvider.provide(tagPairs)
+		if err != nil {
+			panic(err) // Encoding should never fail
+		}
+		rollupIDProvider.finalize()
+		return id
+	}
+
+	// Use default aggregation types, in future we can provide more configurability
+	var defaultAggregationTypes aggregation.TypesConfiguration
+	aggTypeOpts, err := defaultAggregationTypes.NewOptions(instrumentOpts)
+	if err != nil {
+		return newAggregatorResult{}, err
+	}
+
+	ruleSetOpts := rules.NewOptions().
+		SetTagsFilterOptions(tagsFilterOptions).
+		SetNewRollupIDFn(newRollupIDFn).
+		SetIsRollupIDFn(isRollupIDFn).
+		SetAggregationTypesOptions(aggTypeOpts)
+
+	opts := matcher.NewOptions().
+		SetClockOptions(clockOpts).
+		SetInstrumentOptions(instrumentOpts).
+		SetRuleSetOptions(ruleSetOpts).
+		SetKVStore(rulesStore)
+
+	cacheOpts := cache.NewOptions().
+		SetClockOptions(clockOpts).
+		SetInstrumentOptions(instrumentOpts.
+			SetMetricsScope(instrumentOpts.MetricsScope().SubScope("matcher-cache")))
+
+	cache := cache.NewCache(cacheOpts)
+
+	matcher, err := matcher.NewMatcher(cache, opts)
+	if err != nil {
+		return newAggregatorResult{}, err
+	}
+
+	aggregatorOpts := aggregator.NewOptions().
+		SetClockOptions(clockOpts).
+		SetInstrumentOptions(instrumentOpts).
+		SetAggregationTypesOptions(aggTypeOpts).
+		SetMetricPrefix(nil).
+		SetCounterPrefix(nil).
+		SetGaugePrefix(nil).
+		SetTimerPrefix(nil)
+
+	shardSet := make([]shard.Shard, numShards)
```

Comment:
Is the placement always statically configured?

Reply:
Yes, this downsampler always just runs local to the coordinator.